File: 20-actors/mizuho/methods/test_chlorination.py
"""Tests for mizuho residual-dosing (chlorination) operational loop.

    cd 20-actors/mizuho/methods
    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest
"""

from __future__ import annotations

import pytest

from _substrate import PID, SafetyError, simulate
from chlorination import (
    MAX_RESIDUAL_MGL,
    ClampedDoser,
    ResidualChlorinePlant,
    commission_dosing,
    to_datoms,
)


def test_chlorine_holds_target_residual_without_consent():
    # Community-wide disinfection: no per-member consent needed (G6).
    res = commission_dosing(agent="disinfect", target_residual_mgl=0.5)
    assert res.residual_held
    assert res.final_residual_mgl == pytest.approx(0.5, abs=1e-2)
    assert res.ceiling_respected
    assert res.settling_seconds > 0


def test_residual_never_exceeds_regulatory_ceiling():
    # Even commanding a target right at the ceiling, the modeled residual must
    # never cross MAX_RESIDUAL_MGL.
    res = commission_dosing(agent="disinfect", target_residual_mgl=3.9)
    assert res.max_residual_mgl <= MAX_RESIDUAL_MGL + 1e-9
    assert res.ceiling_respected


def test_clamp_holds_even_with_aggressive_gains():
    # The clamp is structural — no choice of gains can drive the residual over
    # the regulatory ceiling.
    plant = ResidualChlorinePlant(residual_mgl=0.0, k_decay=0.0)
    pid = PID(kp=1000.0, ki=1000.0, out_min=0.0, out_max=1e6)
    doser = ClampedDoser(plant, pid, dt=0.1)
    res = simulate(plant, doser, setpoint=999.0, steps=3000, dt=0.1, tol=1e-3)
    worst = max(pv for _, pv, _ in res.trajectory)
    assert worst <= MAX_RESIDUAL_MGL + 1e-9


def test_target_above_ceiling_refused():
    with pytest.raises(SafetyError):
        commission_dosing(agent="disinfect", target_residual_mgl=MAX_RESIDUAL_MGL + 0.1)


def test_fluoride_without_consent_refused_g6():
    # No mandatory fluoridation — anti-paternalism (G6).
    with pytest.raises(SafetyError):
        commission_dosing(agent="fluoridate", target_residual_mgl=0.7)


def test_fluoride_with_consent_passes():
    res = commission_dosing(
        agent="fluoridate", target_residual_mgl=0.7, per_member_consent=True
    )
    assert res.residual_held
    assert res.final_residual_mgl == pytest.approx(0.7, abs=1e-2)
    assert res.ceiling_respected


def test_unknown_agent_refused():
    with pytest.raises(SafetyError):
        commission_dosing(agent="bleach-the-river", target_residual_mgl=0.5)


def test_datoms_are_dry_run_no_server_key():
    res = commission_dosing(agent="disinfect", target_residual_mgl=0.5)
    d = to_datoms(res, "spring-001")
    assert d[":water.dosing/dry-run"] is True
    assert d[":water.dosing/server-held-key"] is False
    assert d[":water.dosing/ceiling-respected"] is True
    assert d[":water.dosing/ceiling-mgl"] == MAX_RESIDUAL_MGL

File: 20-actors/mizuho/methods/test_chlorination.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_chlorination.py.
(ns mizuho.methods.test-chlorination
  "Tests for mizuho residual-dosing (chlorination) operational loop.

  Run:  bb --classpath 20-actors 20-actors/mizuho/methods/test_chlorination.clj"
  (:require [mizuho.methods.substrate :as s]
            [mizuho.methods.chlorination :as ch]
            [clojure.test :refer [deftest is run-tests]]))

(defn- approx [a b tol] (<= (Math/abs (double (- a b))) tol))

(deftest chlorine-holds-target-residual-without-consent
  ;; Community-wide disinfection: no per-member consent needed (G6).
  (let [r (ch/commission-dosing {:agent "disinfect" :target-residual-mgl 0.5})]
    (is (:residual-held r))
    (is (approx (:final-residual-mgl r) 0.5 1e-2))
    (is (:ceiling-respected r))
    (is (> (:settling-seconds r) 0))))

(deftest residual-never-exceeds-regulatory-ceiling
  ;; Even commanding a target right at the ceiling, the residual must never cross it.
  (let [r (ch/commission-dosing {:agent "disinfect" :target-residual-mgl 3.9})]
    (is (<= (:max-residual-mgl r) (+ ch/MAX-RESIDUAL-MGL 1e-9)))
    (is (:ceiling-respected r))))

(deftest clamp-holds-even-with-aggressive-gains
  ;; The clamp is structural — no choice of gains can drive the residual over the ceiling.
  (let [plant (ch/residual-chlorine-plant 0.0 0.0)
        pid (s/make-pid {:kp 1000.0 :ki 1000.0 :out-min 0.0 :out-max 1e6})
        doser (ch/clamped-doser plant pid)
        res (s/simulate plant doser 999.0 3000 0.1 :tol 1e-3)
        worst (reduce max (map second (:trajectory res)))]
    (is (<= worst (+ ch/MAX-RESIDUAL-MGL 1e-9)))))

(deftest target-above-ceiling-refused
  (is (thrown? clojure.lang.ExceptionInfo
               (ch/commission-dosing {:agent "disinfect"
                                      :target-residual-mgl (+ ch/MAX-RESIDUAL-MGL 0.1)}))))

(deftest fluoride-without-consent-refused-g6
  ;; No mandatory fluoridation — anti-paternalism (G6).
  (is (thrown? clojure.lang.ExceptionInfo
               (ch/commission-dosing {:agent "fluoridate" :target-residual-mgl 0.7}))))

(deftest fluoride-with-consent-passes
  (let [r (ch/commission-dosing {:agent "fluoridate" :target-residual-mgl 0.7
                                 :per-member-consent true})]
    (is (:residual-held r))
    (is (approx (:final-residual-mgl r) 0.7 1e-2))
    (is (:ceiling-respected r))))

(deftest unknown-agent-refused
  (is (thrown? clojure.lang.ExceptionInfo
               (ch/commission-dosing {:agent "bleach-the-river" :target-residual-mgl 0.5}))))

(deftest datoms-are-dry-run-no-server-key
  (let [d (ch/to-datoms (ch/commission-dosing {:agent "disinfect" :target-residual-mgl 0.5}) "spring-001")]
    (is (true? (:water.dosing/dry-run d)))
    (is (false? (:water.dosing/server-held-key d)))
    (is (true? (:water.dosing/ceiling-respected d)))
    (is (= (:water.dosing/ceiling-mgl d) ch/MAX-RESIDUAL-MGL))))

;; the G6 refusals are SafetyError-typed (not generic) — assert the type too
(deftest refusals-are-safety-errors
  (doseq [args [{:agent "fluoridate" :target-residual-mgl 0.7}
                {:agent "bleach-the-river" :target-residual-mgl 0.5}
                {:agent "disinfect" :target-residual-mgl 99.0}]]
    (is (try (ch/commission-dosing args) false
             (catch clojure.lang.ExceptionInfo e (s/safety-error? e))))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'mizuho.methods.test-chlorination)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
